### DISCUSSION TITLE
Elementor Form

### DISCUSSION THREAD

**Discussion by advantnet** [elementor/elementor#24049] [Developers Community]
Have any body came across with issue in Elementor form as I am trying to resolve this issue by Disable specific days but is not working, here is the my code:

<script>
  window.addEventListener('DOMContentLoaded', () => {
    const daysOfTheWeek = {
      Monday: 1,
      Tuesday: 2,
      Wednesday: 3,
      Thursday: 4,
      Friday: 5,
      Saturday: 6,
      Sunday: 0
    };
    setTimeout(() => {
      flatpickr('#form-field-field_17eef92', {
        minDate: 'today',
        maxDate: new Date().fp_incr(60), // control the selectable date range
        locale: {
          firstDayOfWeek: daysOfTheWeek['Monday']
        },
        dateFormat: "d-m-Y", // day - month - year format
        disable: [
          function (date) {
            return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday'];
          }
        ],
        ariaDateFormat: 'F j, Y'
      });
    }, 500); // 500ms delay for initializing the date-picker
  });
</script>

would apricate a response

**Comment by perks98** [comment]
Code works for me. Only renders on the frontend and not in the editor. When I tested it I used the code in a HTML widget below the form widget.

![Screenshot_907](https://github.com/elementor/elementor/assets/55829113/9e866fea-2389-401f-a9c8-b09518223299)

**Reply by advantnet** [reply]
Hi perks98

works on Html, but not in Elementor form

here is to show you doesn't work with Elementor popup:

https://www.borgoantico.co.uk/reservations/

Elementor Team say they not support that type of script. what I don't understand is them form

or in any location on the site. The site 100 optimise

Regards

**Reply by perks98** [reply]
Hi @advantnet ,
For me it seems to load the script twice, one on page load (even if popup isn't triggered) and another when the popup is loaded. Very odd. I'm not great with JavaScript but I think I've managed to get it to sort of work, may not be the best way to implement it. Try using the code below.

```
<script>
    var dateFieldUpdater = (function() {
        var executed = false;
        return function() {
            if (!executed) {
                executed = true;
                (async() => {
                    console.log("waiting for flatpickr");
                    while(typeof flatpickr === "undefined") 
                        await new Promise(resolve => setTimeout(resolve, 300));
                    while(document.querySelector("#form-field-field_17eef92") == null) 
                        await new Promise(resolve => setTimeout(resolve, 300));
                    const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 }; 

                    flatpickr('#form-field-field_17eef92', { 
                    minDate: 'today', maxDate: new Date().fp_incr(60), 
                    // control the selectable date range 
                    locale: { firstDayOfWeek: daysOfTheWeek['Monday'] }, 
                    dateFormat: "d-m-Y", 
                    // day - month - year format 
                    disable: [ function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; } ], 
                    ariaDateFormat: 'F j, Y' });
                    
                })();
            }
        };
    })();
    dateFieldUpdater();
</script>
```

**Reply by advantnet** [reply]
Hi Perks98

thanks for your reply, I have tried your script but the issue still the same as you will see from the screenshot when I wrote the script and in the front end go back to normal. to me this is Elementor bug as I have disabled the cache and that have not solved the issue

![booking-issue-backend-231019](https://github.com/elementor/elementor/assets/148346899/4da8b2d6-3b9b-446b-9dbb-67503077921d)


![booking-issue-frontend-231019](https://github.com/elementor/elementor/assets/148346899/e4d79150-1fe2-433c-9de2-689803e195ee)


even I have the the following custom code In Elementor in the header:

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/flatpickr/dist/flatpickr.min.css">

and in the footer:

<script src="https://cdn.jsdelivr.net/npm/flatpickr"></script>

Regards

Phil

**Reply by perks98** [reply]
Hi @advantnet,
Try the code below in your popup, seems to work better. If that doesn't work I don't have a clue.


```
<script>
    jQuery(window).load(function() {
    jQuery.each( elementorFrontend.documentsManager.documents, ( id, documentMod ) => {
			if ( documentMod.getModal ) { 
                documentMod.getModal().on( 'show', () => {
                    const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 };

                    setTimeout(() => { 
					flatpickr("#form-field-field_17eef92", { 
                    minDate: 'today', maxDate: new Date().fp_incr(60), 
                    // control the selectable date range 
                    locale: { firstDayOfWeek: daysOfTheWeek['Monday'] }, 
                    dateFormat: "d-m-Y", 
                    // day - month - year format 
                    disable: [ function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; } ], 
                    ariaDateFormat: 'F j, Y' }); }, 500); 
				} );
			}
		} );
    });
</script>
```

**Reply by advantnet** [reply]
Hi perks98

I tired the new script and did not work, I moved back to the the first one you wrote.

What Elementor sup[port team does not believe what I am saying there is bug in the popup. work on the backend but the front end doesn't.

I hope Elementor Developer Look and take this feedback seriously as I feel Elementor Pro start to bloated with too many scripts.

Kind Regards

**Reply by advantnet** [reply]
Hi Perk98

If you go to the site and trigger the booking first time then close and trigger it aging, You will see the script working.

That is confirm to me that popup have the issue. What is your feedback.

**Reply by perks98** [reply]
Hi @advantnet,
I'm not too sure. For me it works first time I open the popup but when opened again without refreshing the page options go back. I think there is a delay on when the datepicker renders or something.
I think rather than getting the code to run when the popup loads initially it should maybe run based on the interaction with the field. Try this code below in your popup:

```
<script>
    jQuery("#form-field-field_17eef92").one("focus", function(e) {
        const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 };
        flatpickr(this, { 
                    minDate: 'today', maxDate: new Date().fp_incr(60), 
                    // control the selectable date range 
                    locale: { firstDayOfWeek: daysOfTheWeek['Monday'] }, 
                    disableMobile: "true",
                    dateFormat: "d-m-Y", 
                    // day - month - year format 
                    disable: [ function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; } ], 
                    ariaDateFormat: 'F j, Y' });
});
</script>
```

Persoanlly I wouldn't adjust flatpickr datepicker as it doesn't work on iOS, I'd use the native HTML datepicker option and try and adjust that.

**Reply by advantnet** [reply]
Hi perks98

Great news the last code works treat but on on ISO still have to refresh the page.

I wonder how to disable specific dates based on you last code, I want to disable 24-12-2023, 25-12-2023. 01-01-2024, 02-01-2024

👍

**Comment by advantnet** [comment]
Hi perks98 doesn't seam to work on mobile as I tried, also give submission invalid date as you will see from the screenshot
![invalid-date-231020](https://github.com/elementor/elementor/assets/148346899/934f1e6e-8fa0-40b0-bf14-786b925e27f9)
 
I have though working perfect until customer have test it and end submission twice with the same date and time.
is it because you have disable mobile?

**Reply by perks98** [reply]
Hi @advantnet,
Try the code below. From checking online it seems rather than changing the `Format` for flatpickr, you sould use `altFormat` so visually it shows the format you want but saves it how Elementor would normally save it. I think if the date is emailed to your client it may be in the original format rather than the alternative.

```
<script>
    jQuery("#form-field-field_17eef92").one("focus touchstart", function (e) {
    e.preventDefault();
    this.flatpickr().destroy();
    const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 };
    flatpickr(this, {
        minDate: 'today', maxDate: new Date().fp_incr(60),
        // control the selectable date range 
        locale: { firstDayOfWeek: daysOfTheWeek['Monday'] },
        disableMobile: "true",
        altInput: true,
        altFormat: "d-m-Y",
        // day - month - year format 
        disable: [function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; }],
        ariaDateFormat: 'F j, Y'
    });
});
</script>
```

**Reply by perks98** [reply]
Also flatpickr doesn't work well on mobile. Even flatpickr themselves turn it off and change to the default HTML date input when it detects mobile is used.


> "When flatpickr detects a mobile browser, it turns the date input into a native date/time/datetime input."

 -- Quote from https://flatpickr.js.org/mobile-support/

**Reply by advantnet** [reply]
Hi Perks98

I hope you have had a wonderful Christmas.

I am quite surprised as I have tried everything and still showing Monkey and Tuesday on mobile, any idea?
Here is the latest script try to use by help of AI:

<script> jQuery(document).ready(function ($) { const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 }; function initializeFlatpickr() { flatpickr("#form-field-field_17eef92", { minDate: 'today', maxDate: new Date().fp_incr(60), locale: { firstDayOfWeek: daysOfTheWeek['Monday'] }, altInput: true, altFormat: "d-m-Y", disable: [ function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; } ], ariaDateFormat: 'F j, Y' }); } // Initialize Flatpickr $("#form-field-field_17eef92").one("focus touchstart", function (e) { e.preventDefault(); initializeFlatpickr(); }); }); </script>
Looking forward for your reply

**Comment by advantnet** [comment]
Hi Perks98


I hope you have had a wonderful Christmas.

I am quite surprised as I have tried everything and still showing Monkey and Tuesday on mobile, any idea?
Here is the latest script try to use by help of AI:

<script>
jQuery(document).ready(function ($) {
    const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 };

    function initializeFlatpickr() {
        flatpickr("#form-field-field_17eef92", {
            minDate: 'today',
            maxDate: new Date().fp_incr(60),
            locale: { firstDayOfWeek: daysOfTheWeek['Monday'] },
            altInput: true,
            altFormat: "d-m-Y",
            disable: [
                function (date) {
                    return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday'];
                }
            ],
            ariaDateFormat: 'F j, Y'
        });
    }

    // Initialize Flatpickr
    $("#form-field-field_17eef92").one("focus touchstart", function (e) {
        e.preventDefault();
        initializeFlatpickr();
    });
});
</script>

Looking forward for your reply

Regards

**Reply by perks98** [reply]
@advantnet  Try the following code:

```
<script>
    function isInViewport(element) {
        var rect = element.getBoundingClientRect();
        var html = document.documentElement;
        return (
            rect.top >= 0 &&
            rect.left >= 0 &&
            rect.bottom <= (window.innerHeight || html.clientHeight) &&
            rect.right <= (window.innerWidth || html.clientWidth)
        );
    }
    function initializeFlatpickr() {
        const daysOfTheWeek = { Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6, Sunday: 0 };
        flatpickr("#form-field-field_17eef92", {
            minDate: 'today', maxDate: new Date().fp_incr(60),
            locale: { firstDayOfWeek: daysOfTheWeek['Monday'] },
            altInput: true,
            disableMobile: "true",
            altFormat: "d-m-Y",
            disable: [function (date) { return date.getDay() === daysOfTheWeek['Monday'] || date.getDay() === daysOfTheWeek['Tuesday']; }],
            ariaDateFormat: 'F j, Y'
        });
    }
    function runTime() {
        (async () => {
            while (typeof flatpickr === "undefined")
                await new Promise(resolve => setTimeout(resolve, 300));
            while (document.querySelector("#form-field-field_17eef92") == null)
                await new Promise(resolve => setTimeout(resolve, 300));
            while (isInViewport(document.querySelector("#form-field-field_17eef92")) == false)
                await new Promise(resolve => setTimeout(resolve, 300));
            initializeFlatpickr();
        })();
    }

    document.querySelector("#popbut1").addEventListener("click", runTime);
</script>
```
On your 'Book Your Table' button add a CSS ID of 'popbut1', move the HTML widget from the popup to below the button and in the popup settings enable 'Prevent Closing on Overlay'. That then works for me as those options added together trigger a process when you click the button to open the popup it waits until the date input field is in view and when it is it runs the initializeFlatpickr function and works from what I can tell. Having code in the popup is way too messy as it seems to slightly run on page load in some cases and then runs again when you click to load the popup, while this method runs only one on click and then waits until the date field is visible.
